experiments/tf_trainer/tf_kona_prototypical_network/proto.py

```diff
@@ -0,0 +1,317 @@
+    saver.restore(sess, checkpoint)
+    test_itr_handle = sess.run(test_itr.string_handle())
+    while True:
+      try:
+        _, _ = sess.run([update_acc_op, update_auc_op],
+                        feed_dict={handle: test_itr_handle})
+      except tf.errors.OutOfRangeError:
+        break
+    test_acc, test_auc = sess.run([acc_op, auc_op])
+    print("TEST ACCURACY: " + str(test_acc))
+    print("TEST AUC: " + str(test_auc))
+else:
+  print("In TRAINING mode.")
+
+  with tf.gfile.Open(FLAGS.train_file, "r") as f:
+    train_df = pd.read_csv(f)
+    print("Train Dataframe Shape: " + str(train_df.shape))
+    train_dataset = prepare_dataset(train_df).shuffle(128).repeat()
+
+  with tf.gfile.Open(FLAGS.validation_file, "r") as f:
+    validation_df = pd.read_csv(f)
+    print("Validation Dataframe Shape: " + str(validation_df.shape))
+    validation_dataset = prepare_dataset(validation_df).shuffle(64)
+
+  # Training specific model components.
+  with tf.variable_scope("training_operations"):
+    train_op, loss_op = train_operation(negative_logits, positive_logits)
+  with tf.variable_scope("train_predictions_and_metrics"):
+    _, train_acc_op, train_auc_op, train_update_acc_op, train_update_auc_op = predictions_and_metrics(
+        negative_logits, positive_logits)
+  with tf.variable_scope("validation_predictions_and_metrics"):
+    _, val_acc_op, val_auc_op, val_update_acc_op, val_update_auc_op = predictions_and_metrics(
+        negative_logits, positive_logits)
+
+  saver = tf.train.Saver()
+
+  with tf.Session() as sess:
+    sess.run(tf.global_variables_initializer())
+    sess.run(tf.tables_initializer())
+    sess.run(tf.initializers.local_variables())
+
+    train_writer = tf.summary.FileWriter(model_dir + "/train", sess.graph)
+    validation_writer = tf.summary.FileWriter(model_dir + "/validation",
+                                              sess.graph)
+
+    training_iterator = train_dataset.make_one_shot_iterator()
+    validation_iterator = validation_dataset.make_initializable_iterator()
+    training_handle = sess.run(training_iterator.string_handle())
+    validation_handle = sess.run(validation_iterator.string_handle())
+
+    best_auc = 0
+    for batch_num in range(500):
+      print("Batch: " + str(batch_num))
+
+      batch_size = 32
+      for i in range(batch_size):
+        _, loss, train_acc, train_auc = sess.run(
+            [train_op, loss_op, train_update_acc_op, train_update_auc_op],
+            feed_dict={handle: training_handle})
+
+        training_summary = tf.Summary(value=[
+            tf.Summary.Value(tag="loss", simple_value=loss),
+            tf.Summary.Value(tag="accuracy", simple_value=train_acc),
+            tf.Summary.Value(tag="auc", simple_value=train_auc),
+        ])
+        train_writer.add_summary(training_summary, batch_num * batch_size + i)
+        train_writer.flush()
+
+      recent_aucs = collections.deque([], 3)
```

Comment:
Add a comment that this is used for early stopping. I wasn't sure why it was here until I read further in the code.